devito/types/basic.py

```diff
@@ -550,12 +663,50 @@ def name(self):
     @property
     def indices(self):
         """The indices (aka dimensions) of the object."""
-        return self._indices
+        return self.args
+
+    @property
+    def index_ref(self):
```

Comment:
should be called `indices_ref` (or _base ?)

also, this should become an `EnrichedTuple`. I thought I had written about it somewhere, but I lost it. 

```
EnrichedTuple(*self.function.indices, getters=self.dimensions)
```

this way, you can drop `_indices_map`. We use EnrichedTuple in quite a few places (`git grep` to find out), as often we need to index into tuples as if they were dictionaries. If we didn't have EnrichedTuples, we would have tons of additional "dict-ifier" methods

Reply:
this is super important and should be dealt with